nova/objects/console_auth_token.py

```diff
@@ -67,9 +67,19 @@ def access_url(self):
                 # top-level 'token' query parameter was removed. The 'path'
                 # parameter is supported in older noVNC versions, so it is
                 # backward compatible.
-                qparams = {'path': '?token=%s' % self.token}
-                return '%s?%s' % (self.access_url_base,
-                                  urlparse.urlencode(qparams))
+                # NOTE(mariusleu): If the path parameter is already
+                # configured in self.access_url_base, then we should
+                # just append the token to it.
+                url_parts = list(urlparse.urlparse(self.access_url_base))
+                query_params = {}
+                # Transform query_params into simple key-value dictionary
+                for k, v in urlparse.parse_qs(url_parts[4]).items():
+                    query_params[k] = v[0]
+                path = query_params.get('path') or ''
+                path = '%s?token=%s' % (path, self.token)
+                query_params['path'] = path
+                url_parts[4] = urlparse.urlencode(query_params)
```

Comment:
```
>>> o[4] = urlparse.urlencode({'path': 'bla'})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'ParseResult' object does not support item assignment
```
I don't think this works.

There's a `_replace()` method on `ParseResult`, but since it's private, I wouldn't use it. [Here](https://github.com/sapcc/nova/blob/c49cb948bd38ef558c763ddde6db517068579d19/nova/cmd/manage.py#L92-L99) is some code in nova handling an `urlparse`-`urlunparse` workflow. Maybe we should do the same. Instead of building a `ParseResult`, we could also do it like in [oslo.vmware](https://github.com/sapcc/oslo.vmware/blob/af98190afea0a7216b9b79a1f01b5665580c0319/oslo_vmware/rw_handles.py#L433-L443) with a couple of variables.

I would also prefer to use attribute-access instead of the index.

Reply:
`url_parts` is a list (which supports assignment).

Reply:
Ah, right. That's where the list comes from and that's why you can't use speaking names.